### ISSUE TITLE
DSL.value() does not support '-infinity' Postgres timestamp with time zone

### ISSUE THREAD

**Issue by ciaccia** [jOOQ/jOOQ#8027]
Consider the following table:
```sql
CREATE TABLE bug (
   id bigserial PRIMARY KEY,
   last_update timestamp with time zone
);
```

The generated Java class has a custom binder that correctly "parses" '-infinity' timestamps (as in #7126). Here the XML from the configuration:

```xml
<forcedTypes>
	<forcedType>
		<userType>java.time.OffsetDateTime</userType>-->
		<binding>com.example.PGTimestampBinder</binding>
		<expression>.*</expression>
		<types>.*timestamptz.*</types>
	</forcedType>
</forcedTypes>
```
Let's say I want to to write the following SQL query in jOOQ:

```sql
SELECT id, coalesce(last_update, '-infinity'::timestamptz) FROM bug;
```
```java
dsl.select(BUG.ID, DSL.coalesce(BUG.LAST_UPDATE, DSL.value("-infinity", OffsetDateTime.class)))
    .from(BUG)
    .fetch();
````
In the above jOOQ statement null is always passed to coalesce, which is not the expected/desired behaviour.

Furthermore, even if I define my own DataType (as suggested [here](https://stackoverflow.com/questions/51962511/register-global-converters-in-jooq)) I am not able to pass "-infinty" to the underlying DB:

```java
DataType<OffsetDateTime> PG_TIMESTAMP_TYPE = SQLDataType.TIMESTAMPWITHTIMEZONE.asConvertedDataType(new PGTimestampBinder());
dsl.select(BUG.ID, DSL.coalesce(BUG.LAST_UPDATE, DSL.val("-infinity", PG_TIMESTAMP_TYPE)))
    .from(BUG)
    .fetch();
```

To my understanding the problem here seems related to the fact that `DSL.value` tries to convert the passed value immediately, while it should be passed as is to the render() method and pass `-infinity` to the JDBC driver.

Or am I missing something?

**Comment by lukaseder** [comment]
Thanks for your report. Is there anything new compared to your previous report: #7126 ?

**Comment by ciaccia** [comment]
Hi Lukas,
#7126 was about parsing `infinity` and `-infinity` in `org.jooq.impl.DefaultBinding$DelegatingBinding#preparse(String, boolean)` (as suggested in my push request).

This bug is more subtle and a possible fix is not straight clear to me. Let's check that happens in org.jooq.impl.DSL @ line 19800 (version 3.11.2) when calling `DSL.val("-infinity", OffsetDateTime.class)` 

```java
        /* [/pro] */
        // The default behaviour
        else {
            T converted = type.convert(value);
            return new Val<T>(converted, mostSpecific(converted, type));
        }
```

`type.convert("-infinity")` in the third last line returns `null` (i.e. converted is null). This is proably correct since there is no instance of OffsetDateTime which can describe `-infinity`. Then the converted value null is passed to JDBC. 

Theoretically in this case DSL.val(...) should not try to convert `-infinity` at all and store the original value (the string `-infinity` in this case). The original string should then be passed to JDBC when binding the parameters.

**Comment by lukaseder** [comment]
Indeed, your observation here is correct:

> In the above jOOQ statement null is always passed to coalesce, which is not the expected/desired behaviour.

Yes, in the absence of your own data type, jOOQ can only try to auto-convert the `String` to `OffsetDateTime`, and `-infinity` is currently not a recognised value. It should, however be recognised and mapped automatically to `OffsetDateTime.MIN`, which should again be translated to `-infinity` in PostgreSQL. This should be fixed through #7126

> To my understanding the problem here seems related to the fact that DSL.value tries to convert the passed value immediately, while it should be passed as is to the render() method and pass -infinity to the JDBC driver.

The user type `<U>` of your bind value is `OffsetDateTime`, not `String`. So if you're passing an arbitrary String value to the bind value, then jOOQ will need to do something with it first. If you want to encode the PostgreSQL value `-infinity` with your own data type binding, you will need to pass an `OffsetDateTime` value, e.g. `OffsetDateTime.MIN`.

**Comment by lukaseder** [comment]
I will implement #7126 now for jOOQ 3.12. Thanks again for the feedback!
